percona/controllers/pgc/update.go

```diff
@@ -0,0 +1,233 @@
+		strings.ToLower(newCluster.Spec.UpgradeOptions.Apply) == disabled {
+		if ok {
+			c.deleteEnsureVersion(jn)
+		}
+	}
+	if ok && schedule.CronSchedule == newCluster.Spec.UpgradeOptions.Schedule {
+		return nil
+	}
+	if ok {
+		c.deleteEnsureVersion(jn)
+	}
+	id, err := c.crons.crons.AddFunc(newCluster.Spec.UpgradeOptions.Schedule, func() {
+		for i := 1; i < 30; i++ {
+			time.Sleep(5 * time.Second)
+			if !c.crons.inProgress {
+				break
+			}
+		}
+		c.crons.inProgress = true
+		defer func() { c.crons.inProgress = false }()
+
+		err := version.EnsureVersion(newCluster, version.VersionServiceClient{
+			OpVersion: newCluster.ObjectMeta.Labels["pgo-version"],
+		})
+		if err != nil {
+			log.Errorf("schedule update: ensure version %s", err)
+		}
+		oldCluster, err := c.Client.CrunchydataV1().PerconaPGClusters(newCluster.Namespace).Get(context.TODO(), newCluster.Name, metav1.GetOptions{})
+		if err != nil {
+			log.Errorf("get perconapgcluster: %s", err)
+		}
+		err = smartUpdateCluster(c.Client, newCluster, oldCluster)
+		if err != nil {
+			log.Errorf("schedule update: smart update cluster: %s", err)
+		}
+
+		newCluster.ResourceVersion = oldCluster.ResourceVersion
+		_, err = c.Client.CrunchydataV1().PerconaPGClusters(newCluster.Namespace).Update(context.TODO(), newCluster, metav1.UpdateOptions{})
+		if err != nil {
+			log.Errorf("update perconapgcluster: %s", err)
+		}
+	})
+	if err != nil {
+		return errors.Wrap(err, "add func to cron")
+	}
+
+	c.crons.ensureVersionJobs[jn] = Schedule{
+		ID:           int(id),
+		CronSchedule: newCluster.Spec.UpgradeOptions.Schedule,
+	}
+
+	return nil
+}
+
+func smartUpdateCluster(client *kubeapi.Client, newCluster, oldCluster *crv1.PerconaPGCluster) error {
+	log.Printf("Smart update of cluster %s started", oldCluster.Name)
+	log.Printf(`Update with images:
+	%s
+	%s
+	%s
+	%s
+	%s
+	`,
+		newCluster.Spec.Backup.Image,
+		newCluster.Spec.Backup.BackrestRepoImage,
+		newCluster.Spec.PGPrimary.Image,
+		newCluster.Spec.PGBadger.Image,
+		newCluster.Spec.PGBouncer.Image)
+
+	if newCluster.Spec.Backup.Image != oldCluster.Spec.Backup.Image || newCluster.Spec.Backup.BackrestRepoImage != oldCluster.Spec.Backup.BackrestRepoImage {
+		err := restartBackrest(client, oldCluster, newCluster.Spec.Backup.Image, newCluster.Spec.Backup.BackrestRepoImage)
+		if err != nil {
+			return errors.Wrap(err, "restart backrest")
+		}
+	}
+	oldCluster.Spec.Backup.Image = newCluster.Spec.Backup.Image
+	oldCluster.Spec.Backup.BackrestRepoImage = newCluster.Spec.Backup.BackrestRepoImage
+
+	if newCluster.Spec.PGPrimary.Image != oldCluster.Spec.PGPrimary.Image {
+		err := restartCluster(client, oldCluster, newCluster.Spec.PGPrimary.Image)
+		if err != nil {
+			return errors.Wrap(err, "restart cluster")
+		}
+	}
+	oldCluster.Spec.PGPrimary.Image = newCluster.Spec.PGPrimary.Image
+
+	if newCluster.Spec.PGBadger.Image != oldCluster.Spec.PGBadger.Image {
+		err := restartBadger(client, oldCluster, newCluster.Spec.PGBadger.Image)
```

Comment:
Badger it is sidecar for PG nodes. If we have it we need to update and restart it together with PG nodes.

Because now we have two restarts.   Firstly we got wrong image for pgbadger and after the whole reconcile of the pg pods operator detect right image and start to restart pgbadger.

```
  Normal   Pulled                  0s    kubelet                  Container image "perconalab/percona-postgresql-operator:main-ppg13-postgres-ha" already present on machine
  Normal   Created                 0s    kubelet                  Created container database
  Normal   Started                 0s    kubelet                  Started container database
  Normal   Pulled                  0s    kubelet                  Container image "percona/percona-postgresql-operator:1.0.0-ppg13-pgbadger" already present on machine
  Normal   Created                 0s    kubelet                  Created container pgbadger
  Normal   Started                 0s    kubelet                  Started container pgbadger
  Normal   Pulling                 0s    kubelet                  Pulling image "percona/pmm-client:2.23.0
```"